### PR TITLE
mme-realworld

### PR DESCRIPTION
**Pull request by yfzhang114**
MME-RealWorld is the largest manually annotated benchmark to date, featuring the highest resolution and a targeted focus on real-world applications. [https://mme-realworld.github.io/](https://mme-realworld.github.io/)

### CODE REVIEW

**Comment by Luodian** [comment]
Hi thanks for this PR!

File: lmms_eval/tasks/mme_realworld/utils.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        return ''
+    matches = re.search(r'[ABCDE]', s)
+    if matches is None:
+        for choice in choices:
+            if s.lower() in choice.lower():
+                return choice[1]
+        return ''
+    return matches[0]
+
+def mme_doc_to_visual(doc):
```
Comment (kcz358) [inline lmms_eval/tasks/mme_realworld/utils.py]: This function can be deleted

File: lmms_eval/api/task.py [hunk 2]
```diff
@@ -1023,6 +1024,15 @@ def concat_tar_parts(tar_parts, output_tar):
             download_config=download_config,
             **dataset_kwargs if dataset_kwargs is not None else {},
         )
+        
+        selected_datasets = DatasetDict()
+        if 'mmerealworld_cn' in self.task_name:
```
Comment (kcz358) [inline lmms_eval/api/task.py]: Kind of hardcode here for selecting dataset, is it possible that you can split the dataset in different subset or split?

File: lmms_eval/models/internvl2.py [hunk 3]
```diff
@@ -80,7 +80,7 @@ def dynamic_preprocess(image, min_num=1, max_num=6, image_size=448, use_thumbnai
     return processed_images
 
 
-def load_image(image, input_size=448, max_num=6):
+def load_image(image, input_size=448, max_num=25):
```
Comment (kcz358) [inline lmms_eval/models/internvl2.py]: Kindly pin @pufanyi to help check whether the internvl2 should use this settings. Thank you!

**Comment by Luodian** [comment]
and please fix the lint issue.

**Review by kcz358 — CHANGES_REQUESTED** [review]
Hi @yfzhang114 , thank you for your contribution!

I would like to ask a few questions after reviewing the code and kindly ask if some changes can be made before merging.

- The split of dataset currently is hardcoded. This is not a recommended way to do so. Is it possible that you can split your dataset into two splits or two subsets in the huggingface and specify in the yaml?
- As I noticed that the two utils are almost the same instead of the doc_to_text, is it possible that you can merge the two utils into one file? You can specify which doc_to_text to use in the yaml config also.
- This one is not necessary but recommended to do so. I noticed that you convert your dataset from base64 to PIL.Image manually after download. Actually you can upload the dataset in the format of `{"bytes" : <base64 image>, "path" : <a dummy path>}` and set the features to `Image()` from huggingface `dataset` so that the image will show up in your dataset viewer also.

Thank you again for your contribution and please share your thoughts on whether these changes can be made

File: lmms_eval/tasks/mme_realworld/utils.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+#         if "!function" not in line:
+#             safe_data.append(line)
+
+#     config = yaml.safe_load("".join(safe_data))
+
+hf_home = os.getenv("HF_HOME", "~/.cache/huggingface/")
+# cache_dir = os.path.join(hf_home, cache_dir)
+# base_cache_dir = config["dataset_kwargs"]["cache_dir"]
+base_cache_dir = os.path.expanduser(hf_home)
+with open(Path(__file__).parent / "mme_realworld.yaml", "r") as f:
```
Comment (kcz358) [inline lmms_eval/tasks/mme_realworld/utils.py]: If your dataset is purely image and does not rely on gpt-eval, there is no need to load the yaml config from here

File: lmms_eval/models/internvl2.py (hunk 3 above)
Comment (pufanyi) [reply lmms_eval/models/internvl2.py]: Hi, it seems that in the official code of `internvl2`, they use `max_num=6` as default:

https://github.com/OpenGVLab/InternVL/blob/c7c5af1a8930b4862afe8ed14672307082ef61fa/internvl_chat/internvl/train/dataset.py#L702

If you really want to change it, maybe we can change our code to add this parameter to `generation_kwargs`.

Comment (yfzhang114) [reply lmms_eval/models/internvl2.py]: ic, will change it to the default setting

**Comment by yfzhang114** [comment]
> Hi @yfzhang114 , thank you for your contribution!
> 
> I would like to ask a few questions after reviewing the code and kindly ask if some changes can be made before merging.
> 
> * The split of dataset currently is hardcoded. This is not a recommended way to do so. Is it possible that you can split your dataset into two splits or two subsets in the huggingface and specify in the yaml?
> * As I noticed that the two utils are almost the same instead of the doc_to_text, is it possible that you can merge the two utils into one file? You can specify which doc_to_text to use in the yaml config also.
> * This one is not necessary but recommended to do so. I noticed that you convert your dataset from base64 to PIL.Image manually after download. Actually you can upload the dataset in the format of `{"bytes" : <base64 image>, "path" : <a dummy path>}` and set the features to `Image()` from huggingface `dataset` so that the image will show up in your dataset viewer also.
> 
> Thank you again for your contribution and please share your thoughts on whether these changes can be made

1. I separated the English and Chinese versions of MME-RealWorld into two Hugging Face datasets. However, due to the large size of many images, it’s not feasible to upload them in the suggested format {"bytes": <base64 image>, "path": <a dummy path>}.

2. I merged the two utility files into a single file and resolved the lint errors.

File: lmms_eval/api/task.py (hunk 2 above)
Comment (yfzhang114) [reply lmms_eval/api/task.py]: I have split the dataset into different subsets.

**Review by kcz358 — APPROVED** [review]
@yfzhang114 , Thank you for your contribution! LGTM

If it is possible, can you share a screenshot of the evaluation result using the mme-realworld dataset?

And seems like the lint check have failed again. Can you run an auto fixed using pre-commit?

```bash
pip install pre-commit
pre-commit install
pre-commit run --all-files
```

**Comment by yfzhang114** [comment]
@kcz358 of course, here is the results of qwen0.5B_clip_vit, the number of all subtasks has been carefully checked. 

81320073f0125f02d7ce50e9dbfd27c

```
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:149 - ********************************Reasoning (Task Start)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:152 - ++++++++++++++++Monitoring (Subtask Start)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.1867      Calculate (300 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.4592      Intention (98 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.5000      Property (100 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:166 - ++++++++++++++++        Acc 0.3032      E choice 0  Monitoring (498 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:152 - ++++++++++++++++Autonomous_Driving (Subtask Start)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2434      Prediction_intention_ego (304 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.0583      Prediction_intention_pedestrian (103 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2367      Prediction_intention_vehicle (207 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.0547      Relation_interaction_other2other (201 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.4608      Attention_trafficsignal (217 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2264      Relation_interaction_ego2pedestrain (106 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.1238      Relation_interaction_ego2trafficsignal (105 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2376      Relation_interaction_ego2vehicle (101 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:166 - ++++++++++++++++        Acc 0.2240      E choice 0  Autonomous_Driving (1344 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:152 - ++++++++++++++++OCR with Complex Context (Subtask Start)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.4440      Scene understanding (250 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.3880      Character identification (250 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:166 - ++++++++++++++++        Acc 0.4160      E choice 2  OCR with Complex Context (500 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:152 - ++++++++++++++++Diagram and Table (Subtask Start)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2299      Diagram (174 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.3067      Table (326 items)
2024-09-22 03:56:23.257 | INFO     | utils:mme_realworld_aggregate_results:166 - ++++++++++++++++        Acc 0.2800      E choice 0  Diagram and Table (500 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:152 - ++++++++++++++++Remote Sensing (Subtask Start)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:166 - ++++++++++++++++        Acc 0.0000      E choice 0  Remote Sensing (0 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:177 - ********************************Acc 0.2815      E choice 2   Reasoning (2842 items)

2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:149 - ********************************Perception (Task Start)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:152 - ++++++++++++++++Monitoring (Subtask Start)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.1447      Vehicle/counting (608 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2470      Person/counting (992 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.1985      Vehicle/location (136 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.1562      Vehicle/attribute (352 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.1481      Person/attribute (108 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:166 - ++++++++++++++++        Acc 0.1963      E choice 0  Monitoring (2196 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:152 - ++++++++++++++++Autonomous_Driving (Subtask Start)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.3206      Objects_identify (1101 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.4937      Attribute_motion_vehicle/attribute (158 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2612      Attribute_motion_multivehicles/attribute (823 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.3930      Attribute_visual_trafficsignal/attribute (201 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.1951      Attribute_motion_pedestrain/attribute (164 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2042      Object_count (720 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.1785      Attribute_motion_multipedestrians/attribute (493 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:166 - ++++++++++++++++        Acc 0.2710      E choice 0  Autonomous_Driving (3660 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:152 - ++++++++++++++++OCR with Complex Context (Subtask Start)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.3885      License (852 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.3310      Phone_and_address (577 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.4174      Text_recog (1198 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.4204      Adver_and_product (1558 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.4386      Book_map_poster (1555 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:166 - ++++++++++++++++        Acc 0.4110      E choice 0  OCR with Complex Context (5740 items)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:152 - ++++++++++++++++Diagram and Table (Subtask Start)
2024-09-22 03:56:23.258 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2678      Diagram (1415 items)
2024-09-22 03:56:23.259 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2870      Table (4018 items)
2024-09-22 03:56:23.259 | INFO     | utils:mme_realworld_aggregate_results:166 - ++++++++++++++++        Acc 0.2820      E choice 0  Diagram and Table (5433 items)
2024-09-22 03:56:23.259 | INFO     | utils:mme_realworld_aggregate_results:152 - ++++++++++++++++Remote Sensing (Subtask Start)
2024-09-22 03:56:23.259 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2745      Position (1257 items)
2024-09-22 03:56:23.259 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.2677      Color (1255 items)
2024-09-22 03:56:23.259 | INFO     | utils:mme_realworld_aggregate_results:159 - ----   Acc 0.3157      Count (1226 items)
2024-09-22 03:56:23.259 | INFO     | utils:mme_realworld_aggregate_results:166 - ++++++++++++++++        Acc 0.2857      E choice 2  Remote Sensing (3738 items)
2024-09-22 03:56:23.259 | INFO     | utils:mme_realworld_aggregate_results:177 - ********************************Acc 0.3073      E choice 2   Perception (20767 items)

2024-09-22 03:56:23.259 | INFO     | utils:mme_realworld_aggregate_results:178 - ********************************Overall Acc 0.3042
```

**Pull request merged**